discovery/seata-discovery-kubernetes/src/main/java/io/seata/discovery/registry/kubernetes/KubernetesRegistryServiceImpl.java

```diff
@@ -0,0 +1,186 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.discovery.registry.kubernetes;
+
+import io.fabric8.kubernetes.api.model.*;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.seata.common.util.NetUtil;
+import io.seata.config.Configuration;
+import io.seata.config.ConfigurationFactory;
+import io.seata.discovery.registry.RegistryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * @author @hero_zhanghao
+ * @date 2019/5/6 17:55
+ **/
+public class KubernetesRegistryServiceImpl  implements RegistryService<KubernetesEventListener> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KubernetesRegistryServiceImpl.class);
+
+
+    private static volatile boolean subscribeListener = false;
+    private static volatile KubernetesRegistryServiceImpl instance;
+    private static final int MAP_INITIAL_CAPACITY = 8;
+    private static  ConcurrentMap<String, List<InetSocketAddress>> clusterAddressMap;
+
+    private static volatile KubernetesClient kubernetesClient;
+
+    private KubernetesRegistryServiceImpl() {
+    }
+
+
+    static KubernetesRegistryServiceImpl getInstance(KubernetesClient client){
```

Comment:
why need a parameter?

Reply:
> why need a parameter?
Because tests require different implementations, they can also be considered extensions